File: c++/leetcode/merge_two_sorted_lists/merge_two_sorted_lists.cpp
/**
https://leetcode.com/problems/merge-two-sorted-lists/

21. Merge Two Sorted Lists

Merge two sorted linked lists and return it as a new list. The new list should be made by splicing together the nodes of the first two lists.

Example:

	Input: 1->2->4, 1->3->4
	Output: 1->1->2->3->4->4
*/


/**
Solution Outline:
	Start a new empty LL: l3
	Add to the end of l3 choosing from l1 or l2 whichever is smaller
	In the end, append whatever is left of l1 or l2 into l3
*/


#include <iostream>
#include <vector>
#include <functional>
#include <assert.h>
#include <stdio.h>
using namespace std;

/**
 * Definition for singly-linked list.
 */
struct ListNode {
	int val;
	ListNode *next;
	ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
	public:
		ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
			ListNode *merged_ll = new ListNode(0); // dummy head
			ListNode *merged_chain_tail = merged_ll;

			/** Helper lambda to pop head node of a specified linked list */
			// alternately, declare as
			// function<ListNode*(ListNode *)> pop_head = [&](ListNode *ll) -> ListNode* {
			auto pop_head = [&](ListNode *ll) {
				ListNode *p = ll;

				if (ll == l1)
					l1 = l1->next;
				else
					l2 = l2->next;

				p->next = NULL; // sanitize
				return p;
			};

			/** Helper lambda to insert a node/chain to the end of merged chain */
			auto append_to_merged_list = [&](ListNode *chain) {
				merged_chain_tail->next = chain;
				merged_chain_tail = chain;
			};

			while (l1 && l2) {
				ListNode *node;

				if (l1->val < l2->val) {
					node = pop_head(l1);
				} else {
					node = pop_head(l2);
				}
				append_to_merged_list(node);
			}

			/** 
			 * If any of l1 or l2 still has elements left,
			 * then the other list is empty
			 * and everything remaining in this list is > everything seen so far
			 * => just link the remaining chain to the end of the merged list
			 */
			if (l1)
				append_to_merged_list(l1);
			if (l2)
				append_to_merged_list(l2);


			
			/** skip dummy head and return the second node as head of the merged LL */
			return merged_ll->next;
		}
};

/** Convenience function to print a linked list */
void print_ll(struct ListNode *l)
{
	while (l) {
		printf("%d -> ", l->val);

		l = l->next;
	}
	printf("\n");
}


/** Create a linked list from a vector initializer list */
ListNode*
fromList(vector<int> v)
{
	vector<int>::iterator it = v.begin(); 
	ListNode *head = new ListNode(*it);
	ListNode *tail = head;

	while (++it != v.end()) {
		ListNode *node = new ListNode(*it);
		tail->next = node;
		tail = node;
	}

	return head;
}


/** Convenience function to compare linked list elements to a vector */
bool compare_ll_to_vector(ListNode *l, vector<int>&v)
{
	vector<int>::iterator it = v.begin();
	while (l) {
		if(l->val != *it)
			return false;
	
		it++;
		l = l->next;
	}

	return true;
}

int main(void)
{
	vector<int> ev = {1,1,2,3,3,4,5,5};

	ListNode *l1 = fromList(vector<int> {1,2,3,4,5});
	ListNode *l2 = fromList(vector<int> {1,3,5});

	print_ll(l1);
	print_ll(l2);

	Solution s;
	ListNode *merged = s.mergeTwoLists(l1, l2);
	assert(compare_ll_to_vector(merged, ev));

	vector<int> a = {1,2,4};
	vector<int> b = {1,3,4};
	ListNode *la = fromList(vector<int> {1,2,4});

	vector<int> ev2 = {1,1,2,3,4,4};
	assert(compare_ll_to_vector(
			s.mergeTwoLists(
				fromList(vector<int> {1,2,4}), 
				fromList(vector<int> {1,3,4})), 
			ev2));
	return 0;
}

File: python/interviewbit/linked_lists/merge_two_sorted_lists/merge_two_sorted_lists.py
'''
https://www.interviewbit.com/problems/merge-two-sorted-lists/


Merge Two Sorted Lists

Merge two sorted linked lists and return it as a new list.
The new list should be made by splicing together the nodes of the first two lists, and should also be sorted.

For example, given following linked lists :

  5 -> 8 -> 20 
  4 -> 11 -> 15
The merged list should be :

4 -> 5 -> 8 -> 11 -> 15 -> 20
'''


'''
Solution Outline:
	1. Initialize a new linked list chain with a dummy head
		Maintain a tail so appends are efficient.
	2. Pick and yank A's head if A's head <= B's head, otherwise pick and yank B's head
	3. Append yanked node from either A or B into the new chain.
	4. When either of A or B is exhausted, Append the remaining chain to the end of the new chain.
	5. Return new chain after skipping dummy head node as the merged chain.
'''
# Definition for singly-linked list.
class ListNode:
	def __init__(self, x):
		self.val = x
		self.next = None


	# Build a linked-list from a list
	@staticmethod
	def fromList(l):
		head = ListNode(l[0])
		tmp = head
		for x in l[1:]:
			node = ListNode(x)
			tmp.next = node
			tmp = tmp.next
		return head


	# Build a list from a linked-list
	def toList(self):
		tmp = self
		l = []
		while tmp:
			l.append(tmp.val)
			tmp = tmp.next

		return l



class Solution:
    # @param A : head node of linked list
    # @param B : head node of linked list
    # @return the head node in the linked list
    def mergeTwoLists(self, A, B):
        l = ListNode(None) #dummy head
        l_tail = l
        while A and B:
            if A.val <= B.val:
                tmp = A
                A = A.next
            else:
                tmp = B
                B = B.next
            tmp.next = None
            l_tail.next = tmp
            l_tail = tmp
            
        
        if A:
            l_tail.next = A
        elif B:
            l_tail.next = B
            
        return l.next
                

if __name__ == '__main__':
	s = Solution()
	assert s.mergeTwoLists(
						ListNode.fromList([1,3,5,7,9]),
						ListNode.fromList([2,4,6,7,8,9])).toList() == [1,2,3,4,5,6,7,7,8,9,9]
	assert s.mergeTwoLists(
						None,
						ListNode.fromList([1,9])).toList() == [1,9]
	assert s.mergeTwoLists(
						ListNode.fromList([1,9]),
						None).toList() == [1,9]
